File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/IrohaPeerCanonicalTextPayloadCodecV1.java
package org.hyperledger.iroha.android.offline;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Objects;

/** Exact Offline Note UTF-8 application-byte boundary shared by QR, NFC, and Nearby. */
public final class IrohaPeerCanonicalTextPayloadCodecV1 {
  private IrohaPeerCanonicalTextPayloadCodecV1() {}

  public static int maximumCanonicalTextBytes(final IrohaPeerPayloadProfile profile) {
    return maximumCanonicalTextBytes(profile, IrohaPeerWireLimitsV1.PEER_V1);
  }

  public static int maximumCanonicalTextBytes(
      final IrohaPeerPayloadProfile profile, final IrohaPeerWireLimitsV1 limits) {
    Objects.requireNonNull(profile, "profile");
    Objects.requireNonNull(limits, "limits");
    requireOfflineNote(profile);
    return Math.min(limits.maximumCanonicalBytes(), limits.maximumEncodedBytes(profile));
  }

  public static byte[] canonicalBytes(
      final String text, final IrohaPeerPayloadProfile profile) {
    return canonicalBytes(text, profile, IrohaPeerWireLimitsV1.PEER_V1);
  }

  public static byte[] canonicalBytes(
      final String text,
      final IrohaPeerPayloadProfile profile,
      final IrohaPeerWireLimitsV1 limits) {
    Objects.requireNonNull(text, "text");
    requireOfflineNote(profile);
    require(!text.isEmpty(), "Peer canonical text must not be empty");
    final byte[] bytes = text.getBytes(StandardCharsets.UTF_8);
    require(
        new String(bytes, StandardCharsets.UTF_8).equals(text),
        "Peer canonical text is not exact UTF-8");
    requireWithinBound(bytes.length, profile, limits);
    return bytes;
  }

  public static String canonicalText(
      final byte[] bytes, final IrohaPeerPayloadProfile profile) {
    return canonicalText(bytes, profile, IrohaPeerWireLimitsV1.PEER_V1);
  }

  public static String canonicalText(
      final byte[] bytes,
      final IrohaPeerPayloadProfile profile,
      final IrohaPeerWireLimitsV1 limits) {
    Objects.requireNonNull(bytes, "bytes");
    requireOfflineNote(profile);
    require(bytes.length != 0, "Peer canonical text must not be empty");
    requireWithinBound(bytes.length, profile, limits);
    final String text = new String(bytes, StandardCharsets.UTF_8);
    require(
        Arrays.equals(text.getBytes(StandardCharsets.UTF_8), bytes),
        "Peer canonical payload is not exact UTF-8");
    return text;
  }

  private static void requireWithinBound(
      final int actual,
      final IrohaPeerPayloadProfile profile,
      final IrohaPeerWireLimitsV1 limits) {
    final int maximum = maximumCanonicalTextBytes(profile, limits);
    require(
        actual <= maximum,
        "Peer canonical text for " + profile + " is " + actual
            + " bytes; maximum is " + maximum);
  }

  private static void requireOfflineNote(final IrohaPeerPayloadProfile profile) {
    require(
        Objects.requireNonNull(profile, "profile") == IrohaPeerPayloadProfile.OFFLINE_NOTE,
        "Peer generic canonical text is only supported for OFFLINE_NOTE");
  }

  private static void require(final boolean condition, final String message) {
    if (!condition) throw new IllegalArgumentException(message);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/IrohaPeerCanonicalTextPayloadCodecV1.kt
package org.hyperledger.iroha.sdk.offline

/**
 * Exact Offline Note UTF-8 application-byte boundary shared by QR, NFC, and Nearby.
 *
 * Kagemusha profile 2 uses typed native archives and is deliberately rejected
 * here so callers cannot bypass [KagemushaPeerPayload.decode].
 */
object IrohaPeerCanonicalTextPayloadCodecV1 {
    @JvmStatic
    @JvmOverloads
    fun maximumCanonicalTextBytes(
        profile: IrohaPeerPayloadProfile,
        limits: IrohaPeerWireLimitsV1 = IrohaPeerWireLimitsV1.PEER_V1,
    ): Int {
        requireOfflineNote(profile)
        return minOf(
            limits.maximumCanonicalBytes,
            limits.maximumEncodedBytes(profile),
        )
    }

    @JvmStatic
    @JvmOverloads
    fun canonicalBytes(
        text: String,
        profile: IrohaPeerPayloadProfile,
        limits: IrohaPeerWireLimitsV1 = IrohaPeerWireLimitsV1.PEER_V1,
    ): ByteArray {
        requireOfflineNote(profile)
        require(text.isNotEmpty()) { "Peer canonical text must not be empty" }
        val bytes = text.toByteArray(Charsets.UTF_8)
        require(bytes.toString(Charsets.UTF_8) == text) {
            "Peer canonical text is not exact UTF-8"
        }
        requireWithinBound(bytes.size, profile, limits)
        return bytes
    }

    @JvmStatic
    @JvmOverloads
    fun canonicalText(
        bytes: ByteArray,
        profile: IrohaPeerPayloadProfile,
        limits: IrohaPeerWireLimitsV1 = IrohaPeerWireLimitsV1.PEER_V1,
    ): String {
        requireOfflineNote(profile)
        require(bytes.isNotEmpty()) { "Peer canonical text must not be empty" }
        requireWithinBound(bytes.size, profile, limits)
        val text = bytes.toString(Charsets.UTF_8)
        require(text.toByteArray(Charsets.UTF_8).contentEquals(bytes)) {
            "Peer canonical payload is not exact UTF-8"
        }
        return text
    }

    private fun requireWithinBound(
        actual: Int,
        profile: IrohaPeerPayloadProfile,
        limits: IrohaPeerWireLimitsV1,
    ) {
        val maximum = maximumCanonicalTextBytes(profile, limits)
        require(actual <= maximum) {
            "Peer canonical text for ${profile.name} is $actual bytes; maximum is $maximum"
        }
    }

    private fun requireOfflineNote(profile: IrohaPeerPayloadProfile) {
        require(profile == IrohaPeerPayloadProfile.OFFLINE_NOTE) {
            "Peer generic canonical text is only supported for OFFLINE_NOTE"
        }
    }
}
